### PR TITLE
Gaussian prior

### PR DESCRIPTION
**Pull request by NiallJeffrey**
Created a Wiener posterior sampler using the messenger field approach

### CODE REVIEW

**Comment by EiffL** [comment]
@NiallJeffrey sorry we haven't commented on this >.< But know that @b-remy has been looking at it and ran a bunch of comparisons with the other codes. He will comment more here ;-)

**Comment by EiffL** [comment]
@b-remy bump ;-)

**Comment by b-remy** [comment]
Thank you very much @NiallJeffrey for your contribution !

I just added a comparison of your code solution with our method, and as expected our MAP estimator with a gaussian prior is equivalent to your Wiener filter solution ! In another notebook I also show that the MAP estimator is equivalent to the Posterior Mean, so everything seems coherent for the gaussian prior :)

image

**Comment by NiallJeffrey** [comment]
Hi @b-remy - there looks to be some discrepancy. They should be equivalent right? Also, the colour scales are different. Are the power spectra of both methods the same?

**Comment by b-remy** [comment]
Indeed, here is the comparison of the power spectra of both methods and they don't look the same...
image

**Comment by NiallJeffrey** [comment]
This isn't the power spectrum of the Wiener filter/mean of samples. This is the power spectrum of an individual sample, which will have higher power.

The comparison should be the NSM MAP and the Wiener filter

**Comment by NiallJeffrey** [comment]
Are you sure this is the power spectrum of the Wiener filter or the power spectrum of the mean of samples, not the mean power spectrum of the samples?

**Comment by b-remy** [comment]
Yeah I plotted the ps of a sample indeed...
Here is the comparison with the mean of samples:

image

**Comment by b-remy** [comment]
I reviewed your code, and it's looking great to me. I am ready to merge your branch. Are you ok with the comparison now ?

**Comment by NiallJeffrey** [comment]
Yep!

**Comment by NiallJeffrey** [comment]
Can you rerun the comparison using the same input power spectrum in each? The filtered map spectrum looks similar in log scale, but it

**Comment by b-remy** [comment]
Hi @NiallJeffrey , sorry for being so long to answer...

As you proposed, I reran the comparison using the same power spectrum in each. I also added to the repo the corresponding file `jax-lensing/data/ktng_PS_theory.npy`. Results:

image
and
image

So I think it's very coherent now, but I am interested in your opinion.

I also fixed the jit error in `wiener_sampling.py` :-)

**Review by b-remy — APPROVED** [review]
LGTM

File: jax_lensing/wiener_sampling.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        s = ks93inv(s_E,s_B)
+        if i%thinning==0:
+            samples_E[int(i/thinning)] = s_E
+            samples_B[int(i/thinning)] = s_B
+            if verbose==True:
+                print(i)
+    return samples_E, samples_B
+
+
+# doesn't work :/
```
Comment (EiffL) [inline jax_lensing/wiener_sampling.py:122]: so does it work now?

**Comment by EiffL** [comment]
Wow that looks excellent!!! I think it all looks good to merge

Yeah also looks good to me regarding this PR: one question @b-remy it looks like our NSM sampler doesn't quite get the right answer on small scales, is it the same thing

File: jax_lensing/wiener_sampling.py (hunk 1 above)
Comment (b-remy) [reply jax_lensing/wiener_sampling.py]: yes, I just removed this line

**Comment by b-remy** [comment]
I see... Maybe it is because there were not enough samples to compute the mean and there was noise remaining. With a large amount of samples (2e3) small scales agree better.
image

As you can see there is still a little difference when computing the ration so maybe this comes from our sampling procedure. But in this case I don't think it's considerable, is it?

**Comment by NiallJeffrey** [comment]
Have you calculated the auto-correlation length of the MCMC chains? This applies to both this sampling and the general score-learned prior. Generally with these kind of HMC samplers in cosmology (e.g. BORG) the samples are very correlated, so the chains need strong thinning (see Appendix D: https://arxiv.org/pdf/2011.06587.pdf)

**Comment by b-remy** [comment]
In principle samples should be independent because each sample is the last element of a new chain. There is no pair of samples belonging to the same chain.

**Comment by EiffL** [comment]
In any case, I think we can go ahead and merge this branch :-) @b-remy @NiallJeffrey would one of you want to do the honors :-) ?

**Pull request merged**